File: Commons/Refletion/src/main/java/com/greentree/commons/reflection/ClassExtendsTree.java
package com.greentree.commons.reflection;

import java.util.Iterator;

import com.greentree.commons.graph.RootTree;

public final class ClassExtendsTree implements RootTree<Class<?>> {
	
	private static final long serialVersionUID = 1L;
	
	@Override
	public boolean containsInSubTree(Object v, Object parent) {
		return ClassUtil.isExtends((Class<?>) parent, (Class<?>) v);
	}
	
	@Override
	public Class<?> getParent(Object cls) {
		if(cls == Object.class)
			return Object.class;
		return ((Class<?>) cls).getSuperclass();
	}
	
	@Override
	public Iterator<Class<?>> iterator() {
		return ClassUtil.getAllClasses().filter(x -> x.getSuperclass() != null).iterator();
	}
	
}

File: Commons/Reflection/src/main/java/com/greentree/commons/reflection/ClassExtendsTree.kt
package com.greentree.commons.reflection

import com.greentree.commons.graph.RootTree
import com.greentree.commons.reflection.info.TypeUtil

class ClassExtendsTree : RootTree<Class<*>> {

	override fun containsInSubTree(v: Class<*>, parent: Class<*>): Boolean {
		return TypeUtil.isExtends(parent, v)
	}

	override fun getParent(cls: Class<*>): Class<*> {
		if(cls === Any::class.java) return Any::class.java
		return cls.superclass
	}

	override fun iterator(): Iterator<Class<*>> {
		TODO("iterator")
	}
}
